src/app/(authenticated)/chat/[server_name]/[server_id]/[channel_id]/page.tsx

```diff
@@ -0,0 +1,168 @@
+
+const ChannelName = ["General", "Random", "Music", "Gaming", "Anime", "Manga"];
+export default function Page() {
+  const { setChannelId, setServerId } = useContext(WebSocketContext);
+  //contextから引っ張ってきて更新みたいにする
+  //{setchannelId} = context
+  //setchannelId
+  //dynamic paramが変わっても再レンダリングされない
+  const { user } = useUser();
+
+  const router = useRouter();
+  const { server_id, server_name, channel_id } = useParams<Params>();
+  useEffect(() => {
+    console.log("server_id", server_id);
+    console.log("channel_id", channel_id);
+    setChannelId(channel_id);
+    setServerId(server_id);
+  }, [channel_id, server_id]);
+  const [channelInfo, setChannelInfo] = useState<ChannelInfo[] | null>(null);
+
+  const decodedServerName = decodeURIComponent(server_name);
+
+  type InviteToken = {
+    token: string;
+  };
+  const generateInviteTokenAndCopyToClipboard = () => {
+    fetch(
+      `${process.env.NEXT_PUBLIC_BACKEND_ORIGIN}/server/create/invitation`,
+      {
+        method: "POST",
+        headers: {
+          "Content-Type": "application/json",
+        },
+        body: JSON.stringify({
+          user_id: user?.id,
+          server_id: server_id,
+        }),
+      }
+    )
+      .then(async (response) => {
+        if (response.ok) {
+          console.log("token generate success!");
+          const inviteToken: InviteToken = await response.json();
+          await navigator.clipboard.writeText(inviteToken.token);
+          toast({
+            title: "Invite token copyed!",
+          });
+        } else {
+          const message = await response.text();
+          throw Error(message);
+        }
+      })
+      .catch((error) => {
+        console.error(
+          "error occured while generate invite server tokne:",
+          error
+        );
+      });
+  };
+
+  //websocketで情報を受け取ってstateの配列に追加
+  //channel_idなどをuseEffectの第二引数に渡してdynamic paramが変更されてチャンネル
+  //が切り替わったときにcontextに保持している配列を初期化するような
+  //hooksを実行させる
+  useEffect(() => {
+    console.log("server name is ", server_name);
+    const controller = new AbortController();
+    const fetchChannelInfo = async () => {
+      try {
+        const response = await fetch(
+          `${process.env.NEXT_PUBLIC_BACKEND_ORIGIN}/channels/${server_id}`,
+          {
+            cache: "no-store",
+            signal: controller.signal,
+          }
+        );
+        if (!response.ok) {
+          throw new Error("Failed to fetch channel info");
+        }
+        console.log("fetch channel list is success!");
+        const initChannelInfo: ChannelInfo[] = await response.json();
+        setChannelInfo(initChannelInfo);
+
+        console.log("channelInfo:", channelInfo);
+        if (
+          !initChannelInfo?.some((channel) => channel.channel_id === channel_id)
+        ) {
+          if (initChannelInfo?.[0]) {
+            router.push(
+              `/chat/${server_name}/${server_id}/${initChannelInfo[0].channel_id}`
+            );
+          }
+        }
```

Comment:
if文のネストは減らしたいから早期リターンすることを意識するといいかも
```suggestion
        if (initChannelInfo?.some((channel) => channel.channel_id === channel_id)) {
           return;
        }
        router.push(
              `/chat/${server_name}/${server_id}/${initChannelInfo[0].channel_id}`
        );
```

` if (initChannelInfo?.[0]) {`ここのif文は必要なのか🧐

Reply:
initChannelInfoにfetchされた結果として[]が入った場合`/chat/${server_name}/${server_id}/`にrouter.pushが行われることになるので、チャンネル情報自体が配列に入っているかを確認するためにif文を書いています

Reply:
initChannelInfoの中にchannel_idと一致するものがある場合router.push()するでも良いと思う
もしinitChannelInfoが空だった場合、initChannelInfo?.some((channel) => channel.channel_id === channel_id)はfalseを返すはず